### PR TITLE
Implement Bot API 6.5

### PR DESCRIPTION
**Pull request by tuscen**

### CODE REVIEW

File: src/Telegram.Bot/Types/Message.cs [hunk 1]
```diff
@@ -377,6 +377,18 @@ public class Message
     [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
     public SuccessfulPayment? SuccessfulPayment { get; set; }
 
+    /// <summary>
+    /// Optional. Service message: a user was shared with the bot
+    /// </summary>
+    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+    public UserShared UserShared { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/Message.cs:384]:
```suggestion
    public UserShared? UserShared { get; set; }
```

File: src/Telegram.Bot/Types/Message.cs [hunk 2]
```diff
@@ -377,6 +377,18 @@ public class Message
...
+
+    /// <summary>
+    /// Optional. Service message: a chat was shared with the bot
+    /// </summary>
+    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+    public ChatShared ChatShared { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/Message.cs:390]:
```suggestion
    public ChatShared? ChatShared { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /// <summary>
+    /// Signed 32-bit identifier of the request
+    /// </summary>
+    public int RequestId { get; set; }
+
+    /// <summary>
+    /// Pass <see langword="true" /> to request a channel chat, pass <see langword="false" /> to request a group
+    /// or a supergroup chat.
+    /// </summary>
+    public bool? ChatIsChannel { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:20]:
```suggestion
    public bool ChatIsChannel { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+/// shared with the bot when the corresponding button is pressed.
+/// </summary>
+[JsonObject(MemberSerialization.OptIn, NamingStrategyType = typeof(SnakeCaseNamingStrategy))]
+public class KeyboardButtonRequestChat
+{
+
+    /// <summary>
+    /// Signed 32-bit identifier of the request
+    /// </summary>
+    public int RequestId { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:14]:
```suggestion
    [JsonProperty(Required = Required.Always)]
    public int RequestId { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs (hunk 3 above)
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:20]:
```suggestion
   [JsonProperty(Required = Required.Always)]
    public bool? ChatIsChannel { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    /// Pass <see langword="true" /> to request a channel chat, pass <see langword="false" /> to request a group
+    /// or a supergroup chat.
+    /// </summary>
+    public bool? ChatIsChannel { get; set; }
+
+    /// <summary>
+    /// Optional. Pass <see langword="true" /> to request a forum supergroup, pass <see langword="false" /> to
+    /// request a non-forum chat. If not specified, no additional restrictions are applied.
+    /// </summary>
+    public bool? ChatIsForum { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:26]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public bool? ChatIsForum { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Optional. Pass <see langword="true" /> to request a supergroup or a channel with a username,
+    /// pass <see langword="false" /> to request a chat without a username. If not specified, no additional
+    /// restrictions are applied.
+    /// </summary>
+    public bool? ChatHasUsername { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:33]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public bool? ChatHasUsername { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Optional. Pass <see langword="true" /> to request a chat owned by the user. Otherwise, no additional
+    /// restrictions are applied.
+    /// </summary>
+    public bool? ChatIsCreated { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:39]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public bool? ChatIsCreated { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Optional. A JSON-serialized object listing the required administrator rights of the user in the chat.
+    /// If not specified, no additional restrictions are applied.
+    /// </summary>
+    public ChatAdministratorRights? UserAdministratorRights { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:45]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public ChatAdministratorRights? UserAdministratorRights { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Optional. A JSON-serialized object listing the required administrator rights of the bot in the chat.
+    /// The rights must be a subset of user_administrator_rights. If not specified, no additional
+    /// restrictions are applied.
+    /// </summary>
+    public ChatAdministratorRights? BotAdministratorRights { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:52]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public ChatAdministratorRights? BotAdministratorRights { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// Optional. Pass <see langword="true" /> to request a chat with the bot as a member. Otherwise, no additional
+    /// restrictions are applied.
+    public bool BotIsMember { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:56]:
```suggestion
    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
    public bool BotIsMember { get; set; }
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestUser.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+public class KeyboardButtonRequestUser
+{
+    /// <summary>
+    /// Signed 32-bit identifier of the request
+    /// </summary>
+    [JsonProperty(Required = Required.Always)]
+    public int RequestId { get; set; }
+
+    /// <summary>
+    /// Optional. Pass True to request a bot, pass False to request a regular user. If not specified, no additional
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestUser.cs:17]:
```suggestion
    /// Optional. Pass <see langword="true" /> to request a bot, pass <see langword="false" /> to request a regular user. If not specified, no additional
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestUser.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    /// restrictions are applied.
+    /// </summary>
+    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+    public bool? UserIsBot { get; set; }
+
+    /// <summary>
+    /// Optional. Pass True to request a premium user, pass False to request a non-premium user. If not specified,
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestUser.cs:24]:
```suggestion
    /// Optional. Pass <see langword="true" /> to request a premium user, pass <see langword="false" /> to request a non-premium user. If not specified,
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// <summary>
+    /// Optional. A JSON-serialized object listing the required administrator rights of the user in the chat.
+    /// If not specified, no additional restrictions are applied.
+    /// </summary>
+    public ChatAdministratorRights? UserAdministratorRights { get; set; }
+
+    /// <summary>
+    /// Optional. A JSON-serialized object listing the required administrator rights of the bot in the chat.
+    /// The rights must be a subset of user_administrator_rights. If not specified, no additional
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButtonRequestChat.cs:49]: I'm not sure this reference is correct:
```suggestion
    /// The rights must be a subset of <see cref="ChatAdministratorRights" />. If not specified, no additional
```

File: src/Telegram.Bot/Types/UserShared.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
+using Telegram.Bot.Types.ReplyMarkups;
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/UserShared.cs:1]: This type should be in the location with ChatShare.cs, isn't it?

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButton.cs [hunk 15]
```diff
@@ -26,8 +26,22 @@ public class KeyboardButton : IKeyboardButton
     public string Text { get; set; }
 
     /// <summary>
-    /// Optional. If <see langword="true"/>, the user's phone number will be sent as a contact when the button is pressed.
-    /// Available in private chats only
+    /// Optional. If specified, pressing the button will open a list of suitable users. Tapping on any user will send
+    /// their identifier to the bot in a “user_shared” service message. Available in private chats only.
+    /// </summary>
+    [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+    public KeyboardButtonRequestUser? RequestUser { get; set; }
```
Comment (karb0f0s) [inline src/Telegram.Bot/Types/ReplyMarkups/KeyboardButton.cs:33]: should we add static helpers for new button types?

File: CHANGELOG.md [hunk 16]
```diff
@@ -21,7 +21,33 @@ and this project adheres to [Semantic Versioning](https://semver.org/).
...
 
-## [v19.0.0-preview.2] - Unreleased
+## [v19.0.0-preview.3] - Unreleased
+
+> [Bot API 6.5](https://core.telegram.org/bots/api#february-3-2023) (February 3, 2023)
+
+### Added
+
+- Types `KeyboardButtonRequestUser` and `KeyboardButtonRequestChat`
+- Properties `RequestUser` and `RequestUser` to the type `KeyboardButton`
```
Comment (karb0f0s) [inline CHANGELOG.md:31]:
```suggestion
- Properties `RequestUser` and `RequestChat` to the type `KeyboardButton`
```

File: src/Telegram.Bot/Types/ReplyMarkups/KeyboardButton.cs (hunk 15 above)
Comment (tuscen) [reply src/Telegram.Bot/Types/ReplyMarkups/KeyboardButton.cs:33]: Done

**Review by karb0f0s — APPROVED** [review]

**Review by karb0f0s — APPROVED** [review]

**Pull request merged**